File: Python/algoritms/sort/merge_sort.py


def merge(left_half: list, right_half: list) -> list:
    sorted_array = [None] * (len(right_half) + len(left_half))

    pointer1 = 0  # pointer to current index for left Half
    pointer2 = 0  # pointer to current index for the right Half
    index = 0  # pointer to current index for the sorted array Half

    while pointer1 < len(left_half) and pointer2 < len(right_half):
        if left_half[pointer1] < right_half[pointer2]:
            sorted_array[index] = left_half[pointer1]
            pointer1 += 1
            index += 1
        else:
            sorted_array[index] = right_half[pointer2]
            pointer2 += 1
            index += 1
    while pointer1 < len(left_half):
        sorted_array[index] = left_half[pointer1]
        pointer1 += 1
        index += 1

    while pointer2 < len(right_half):
        sorted_array[index] = right_half[pointer2]
        pointer2 += 1
        index += 1

    return sorted_array

def merge_sort(array):
    if len(array) <= 1:
        return array
    # the actual formula to calculate the middle element = left + (right - left) // 2
    # this avoids integer overflow in case of large N
    middle = 0 + (len(array) - 0) // 2

    # Split the array into halves till the array length becomes equal to One
    # merge the arrays of single length returned by mergeSort function and
    # pass them into the merge arrays function which merges the array
    left_half = array[:middle]
    right_half = array[middle:]

    return merge(merge_sort(left_half), merge_sort(right_half))

print(merge_sort([1,3,2,5,7,6]))

File: Go/sort/mergeSort.go
package sort

import (
	"sync"
)

func Mergesort(array []int) []int {
	if len(array) <= 1 {
		return array
	}

	mid := len(array) / 2

	leftHalf := array[:mid]
	rightHalf := array[mid:]

	return merge(Mergesort(leftHalf), Mergesort(rightHalf))

}

func merge(i1, i2 []int) []int {
	size := len(i1) + len(i2)
	sortedArray := make([]int, size)

	pointer1 := 0
	pointer2 := 0
	index := 0

	for pointer1 < len(i1) && pointer2 < len(i2) {
		if i1[pointer1] < i2[pointer2] {
			sortedArray[index] = i1[pointer1]
			pointer1++
			index++
		} else {
			sortedArray[index] = i2[pointer2]
			pointer2++
			index++
		}
	}
	for pointer1 < len(i1) {
		sortedArray[index] = i1[pointer1]
		pointer1++
		index++
	}

	for pointer2 < len(i2) {
		sortedArray[index] = i2[pointer2]
		pointer2++
		index++
	}
	return sortedArray
}

func Merge(items []int) []int {

	if len(items) < 2 {
		return items

	}

	var middle = len(items) / 2
	var a = Merge(items[:middle])
	var b = Merge(items[middle:])
	return merge(a, b)

}

// ParallelMerge Perform merge sort on a slice using goroutines
func ParallelMerge(items []int) []int {
	if len(items) < 2 {
		return items
	}

	if len(items) < 2048 {
		return Merge(items)
	}

	var wg sync.WaitGroup
	wg.Add(1)

	var middle = len(items) / 2
	var a []int
	go func() {
		defer wg.Done()
		a = ParallelMerge(items[:middle])
	}()
	var b = ParallelMerge(items[middle:])

	wg.Wait()
	return merge(a, b)
}
